### ISSUE TITLE
Names with shared generics all use one unrelated name in ida.py?

### ISSUE THREAD

**Issue by DannyParker0001** [djkaty/Il2CppInspector#53]
I'm pretty sure it has to do with generic sharing? Not too sure what to call it. I'm assuming this is a bug...  
In Ida, I noticed I kept seeing a reference to: `Dictionary_2_System_Int32_UnityEngine_Experimental_TerrainAPI_TerrainUtility_TerrainMap_set_Item`, I checked xrefs to the func and it has 756 references, mostly from completely unrelated methods. I looked at other `Dictionary_2_System_Int32_UnityEngine_Experimental_TerrainAPI_TerrainUtility_TerrainMap_xxxx` methods and a lot of them have >100 xrefs. 
I'm going to assume its meant to be `Dictionary_System_Int32_XXXXXX_set_Item` rather than just one name?
[Game files here if you need them](https://github.com/djkaty/Il2CppInspector/files/4947553/Files.zip)

**Comment by djkaty** [comment]
Thanks Danny,

Yeah we don't do anything about generic sharing right now; since you can't set multiple labels for one address, whichever is the last one in ida.py will end up being the final name. If you grep ida.py for `1811357C0` in your game you'll see all the names, with the Terrain one being the last one defined.

I've added a note in my TODOs to look at this later but I'll be treating t his as a low priority item for now until some other unfinished major features are completed. In the meantime it's probably easiest to just rename the addresses that are problematic in IDA.

@nneonneo is our resident generic expert - perhaps you have an idea on how to handle this? :)

**Comment by nneonneo** [comment]
This was an item I wanted to fix in the generics update, but didn’t get around to it. As far as I can tell, during linking, a typical linker will merge functions with identical bodies, even if the functions have different signatures - this is called ICF (identical code folding). Many Il2Cpp-generated functions are highly similar, often resulting in multiple functions being merged into one - even completely unrelated ones. 

My plan was to basically go through any binary function that corresponds to multiple methods, extracting the least common denominator among the parameters and return type (e.g. the common base class), and identify a possible common name for the function. When I get some time I will work on a patch to do exactly this.

**Comment by djkaty** [comment]
Agreed, that was what I figured as well. Although IL2CPP specifically merges functions via generic sharing besides classical C++ linker optimizations. Either way the resultant problem is the same I believe.

I'll assign this to you @nneonneo to just do whenever you feel like it.

There is a wall of commits going up soon which significantly changes the behaviour of the UnityHeader code and some more minor changes to CppDeclarationGenerator, plus an AppModel with composite types as we discussed earlier, which means the way the C++ and IDA output modules fetch data has completely changed and all routes via AppModel now.

Therefore, in order to implement this, we are going to need to add a public collection property to AddModel I think which groups all symbols for a specific virtual address together, then the output module can decide what to do with them.

**Comment by djkaty** [comment]
Besides building an in-memory address map of items in AppModel, all of the changes to the implementation of modelling the binary and how this affects output modules is now live, so feel free to look over it and see if you can find an elegant way to deal with the generic sharing / ICF problem if you like!

**Comment by lifeengines** [comment]
> This was an item I wanted to fix in the generics update, but didn’t get around to it. As far as I can tell, during linking, a typical linker will merge functions with identical bodies, even if the functions have different signatures - this is called ICF (identical code folding). Many Il2Cpp-generated functions are highly similar, often resulting in multiple functions being merged into one - even completely unrelated ones. This is most visible with very simple functions (e.g. no-op functions or basic getters).
> 
> My plan was to basically go through any binary function that corresponds to multiple methods, extracting the least common denominator among the parameters and return type (e.g. the common base class), and identify a possible common name for the function. When I get some time I will work on a patch to do exactly this.

that's too complicated,most of these method has MethodInfo parmater. this would determine which function it real is.
